keps/sig-node/2133-kubelet-credential-providers/README.md

```diff
@@ -0,0 +1,517 @@
+- [X] (R) KEP approvers have approved the KEP status as `implementable`
+- [X] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [X] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubelet has a credential provider mechanism, which gives kubelet the ability to dynamically
+fetch credentials for image registries. Today there are three built-in implementations of the
+kubelet credential provider for ACR (Azure Container Registry), ECR (Elastic Container Registry),
+and GCR (Google Container Registry). This KEP proposes an extensible plugin mechanism so that
+kubelet can dynamically fetch image registry credentials for any cloud provider.
+
+## Motivation
+
+This is part of a larger effort in the project to remove built-in functionality that favors any
+specific cloud provider. The ACR, ECR and GCR implementations should be removed in favor of an
+extensible plugin mechanism that can be used by any cloud provider without adding built-in logic
+into the kubelet.
+
+### Goals
+
+* add a plugin mechanism so that kubelet can dynamically fetch credentials for a given registry.
+* the plugin should have feature parity with existing functionality in-tree.
+
+### Non-Goals
+
+* removing the built-in ACR, ECR and GCR implementations in kubelet. This is an end-goal but not a goal of this KEP.
+* improving image pull performance of kubelet.
+* improving kubelet security around using image pull credentials.
+
+## Proposal
+
+The extension mechanism introduced in the kubelet will be done by exec-ing a plugin binary.
+The kubelet and the plugin communicate through stdio (stdin, stdout, and stderr) by sending
+and receiving json-serialized api-versioned types. The kubelet and the plugin must always
+talk the same api version to ensure compatibility as the API evolves.
+
+### Risks and Mitigations
+
+* in contrast to existing built-in implementations, credentials for a image registry is now passed
+through stdio of a process invoked by the kubelet, as opposed to those credentials only remaining in-memory.
+* exec-ing plugins for image credentials can be expensive for the kubelet.
```

Comment:
Might this be an appropriate place to mention the strategy that the current implementation takes of timing out plugins after 1 minute to prevent plugin processes from becoming too long-lived? LGTM otherwise, I think.

Reply:
Good catch! Will make sure to include that for the v1.21 beta update of this KEP :)